diff --git a/components/freertos/include/freertos/FreeRTOSConfig.h b/components/freertos/include/freertos/FreeRTOSConfig.h
index 3ccd3fd..6c972dc 100644
--- a/components/freertos/include/freertos/FreeRTOSConfig.h
+++ b/components/freertos/include/freertos/FreeRTOSConfig.h
@@ -190,7 +190,7 @@
 #define configTOTAL_HEAP_SIZE                  (&_heap_end - &_heap_start)//( ( size_t ) (64 * 1024) )

 #define configMAX_TASK_NAME_LEN                        ( CONFIG_FREERTOS_MAX_TASK_NAME_LEN )
-#define configUSE_TRACE_FACILITY               0               /* Used by vTaskList in main.c */
+#define configUSE_TRACE_FACILITY               1 /* Used by vTaskList in main.c */
 #define configUSE_STATS_FORMATTING_FUNCTIONS   0       /* Used by vTaskList in main.c */
 #define configUSE_TRACE_FACILITY_2      0              /* Provided by Xtensa port patch */
 #define configBENCHMARK                                        0               /* Provided by Xtensa port patch */
