diff --git a/components/lwip/port/freertos/sys_arch.c b/components/lwip/port/freertos/sys_arch.c
index aa58f9bb..97ee4216 100755
--- a/components/lwip/port/freertos/sys_arch.c
+++ b/components/lwip/port/freertos/sys_arch.c
@@ -441,7 +441,7 @@ sys_init(void)
     // Create the pthreads key for the per-thread semaphore storage
     pthread_key_create(&sys_thread_sem_key, sys_thread_sem_free);
 
-    esp_vfs_lwip_sockets_register();
+    // esp_vfs_lwip_sockets_register();
 }
 
 /*-----------------------------------------------------------------------------------*/
