diff --git a/components/lwip/port/freertos/sys_arch.c b/components/lwip/port/freertos/sys_arch.c
index 6f14a3a..ec07c15 100755
--- a/components/lwip/port/freertos/sys_arch.c
+++ b/components/lwip/port/freertos/sys_arch.c
@@ -433,7 +433,7 @@ sys_init(void)
     if (ERR_OK != sys_mutex_new(&g_lwip_protect_mutex)) {
         ESP_LOGE(TAG, "sys_init: failed to init lwip protect mutex\n");
     }
-    esp_vfs_lwip_sockets_register();
+    //esp_vfs_lwip_sockets_register();
 }
 
 /*-----------------------------------------------------------------------------------*/
