diff --git a/components/vfs/include/esp_vfs.h b/components/vfs/include/esp_vfs.h
index ab645fcc..c63f83cd 100644
--- a/components/vfs/include/esp_vfs.h
+++ b/components/vfs/include/esp_vfs.h
@@ -15,6 +15,8 @@
 #ifndef __ESP_VFS_H__
 #define __ESP_VFS_H__
 
+#include "luartos.h"
+
 #include <stdint.h>
 #include <stddef.h>
 #include <stdarg.h>
@@ -24,6 +26,8 @@
 #include <sys/stat.h>
 #include <dirent.h>
 
+#include "lwip/sockets.h"
+
 #ifdef __cplusplus
 extern "C" {
 #endif
@@ -166,6 +170,14 @@ typedef struct
         int (*fsync_p)(void* ctx, int fd);
         int (*fsync)(int fd);
     };
+    union {
+        int (*writev_p)(void* ctx, int fd, const struct iovec *iov, int iovcnt);
+        int (*writev)(int fd, const struct iovec *iov, int iovcnt);
+    };
+    union {
+        int (*select_p)(void* ctx, int maxfdp1, fd_set *readset, fd_set *writeset, fd_set *exceptset, struct timeval *timeout);
+        int (*select)(int maxfdp1, fd_set *readset, fd_set *writeset, fd_set *exceptset, struct timeval *timeout);
+    };
 } esp_vfs_t;
 
 
diff --git a/components/vfs/vfs.c b/components/vfs/vfs.c
index 5dbd462c..28f9c96c 100644
--- a/components/vfs/vfs.c
+++ b/components/vfs/vfs.c
@@ -40,7 +40,7 @@
 // mask of actual file descriptor (e.g. 0x00000fff)
 #define VFS_FD_MASK     ((1 << CONFIG_MAX_FD_BITS) - 1)
 // max number of VFS entries
-#define VFS_MAX_COUNT   ((1 << MAX_VFS_ID_BITS) - 1)
+// #define VFS_MAX_COUNT   ((1 << MAX_VFS_ID_BITS) - 1)
 // mask of VFS id (e.g. 0x00007000)
 #define VFS_INDEX_MASK  (VFS_MAX_COUNT << CONFIG_MAX_FD_BITS)
 #define VFS_INDEX_S     CONFIG_MAX_FD_BITS
@@ -269,6 +269,7 @@ ssize_t esp_vfs_write(struct _reent *r, int fd, const void * data, size_t size)
         __errno_r(r) = EBADF;
         return -1;
     }
+
     int local_fd = translate_fd(vfs, fd);
     ssize_t ret;
     CHECK_AND_CALL(ret, r, vfs, write, local_fd, data, size);
@@ -552,3 +553,191 @@ int fsync(int fd)
     CHECK_AND_CALL(ret, r, vfs, fsync, local_fd);
     return ret;
 }
+
+ssize_t writev(int fd, const struct iovec *iov, int iovcnt)
+{
+    const vfs_entry_t* vfs = get_vfs_for_fd(fd);
+    struct _reent* r = __getreent();
+    if (vfs == NULL) {
+        __errno_r(r) = EBADF;
+        return -1;
+    }
+    int local_fd = translate_fd(vfs, fd);
+    ssize_t ret;
+    CHECK_AND_CALL(ret, r, vfs, writev, local_fd, iov, iovcnt);
+    return ret;
+}
+
+#define VFS_FD_SET_NUM 3
+
+typedef struct {
+   fd_set readset;
+   fd_set writeset;
+   fd_set exceptset;
+
+   fd_set test_readset;
+   fd_set test_writeset;
+   fd_set test_exceptset;
+
+   const vfs_entry_t *vfs;
+} vfs_fd_set_t;
+
+static vfs_fd_set_t *get_vfs_fd_set_for(int fd, vfs_fd_set_t *vfs_fd_set) {
+   const vfs_entry_t *vfs = get_vfs_for_fd(fd);
+    if (vfs) {
+       int min_fd = vfs->offset << VFS_INDEX_S;
+       int max_fd = (vfs->offset + 1) << VFS_INDEX_S;
+
+       if ((fd >= min_fd) && (fd < max_fd)) {
+           int i;
+
+           // Search if vfs has an assigned fd_set
+           for(i = 0;i < VFS_FD_SET_NUM;i++) {
+               if (vfs_fd_set[i].vfs == vfs) {
+                   return &vfs_fd_set[i];
+               }
+           }
+
+           // Find an empty fd_set fot the vfs
+           for(i = 0;i < VFS_FD_SET_NUM;i++) {
+               if (!vfs_fd_set[i].vfs) {
+                   vfs_fd_set[i].vfs = vfs;
+
+                   return &vfs_fd_set[i];
+               }
+           }
+       }
+    }
+
+   return NULL;
+}
+
+int __select_cancelled = 0;
+
+int vfs_select (int maxfdp1, fd_set *readset, fd_set *writeset, fd_set *exceptset, struct timeval *timeout) {
+   // Convert the timeout to milliseconds
+   u32_t msectimeout = 0;
+
+   if (timeout && (timeout->tv_sec != 0) && (timeout->tv_usec != 0)) {
+      msectimeout = ((timeout->tv_sec * 1000) + ((timeout->tv_usec + 500)/1000));
+   }
+
+   // Initialize fd_set for each vfs
+   vfs_fd_set_t vfs_fd_set[VFS_FD_SET_NUM];
+
+   int i;
+
+   for(i = 0;i < VFS_FD_SET_NUM;i++) {
+       FD_ZERO(&vfs_fd_set[i].readset);
+       FD_ZERO(&vfs_fd_set[i].writeset);
+       FD_ZERO(&vfs_fd_set[i].exceptset);
+       vfs_fd_set[i].vfs = NULL;
+   }
+
+   // Split readset / writeset / exceptset to individuals fd_set
+   // for each involved file system
+   vfs_fd_set_t *fd_sets;
+
+   int fd;
+
+   for(fd = 0;fd <= maxfdp1;fd++) {
+       if (readset && FD_ISSET(fd, readset)) {
+           fd_sets = get_vfs_fd_set_for(fd, vfs_fd_set);
+           if (fd_sets) {
+               FD_SET(fd, &fd_sets->readset);
+           }
+       }
+
+       if (writeset && FD_ISSET(fd, writeset)) {
+           fd_sets = get_vfs_fd_set_for(fd, vfs_fd_set);
+           if (fd_sets) {
+               FD_SET(fd, &fd_sets->writeset);
+           }
+       }
+
+       if (exceptset && FD_ISSET(fd, exceptset)) {
+           fd_sets = get_vfs_fd_set_for(fd, vfs_fd_set);
+           if (fd_sets) {
+               FD_SET(fd, &fd_sets->exceptset);
+           }
+       }
+   }
+
+   // Call select for each involved file system
+   struct timeval fs_timeout = {0, 1000}; // Set a timeout of 1 millisecond for each fs
+   struct timeval start; // Start time
+   struct timeval now;   // Current time
+
+   int num = 0; // Number of set file descriptors
+   int min_fd;  // Min file descriptor in the current file system
+   int max_fd;  // Max file descriptor in the current file system
+
+   // Get start time
+   gettimeofday(&start, NULL);
+
+   // Scan file descriptors for each involved file system until a file descriptor is set or
+   // until timeout
+   struct _reent* r = __getreent();
+   int ret;
+
+   for(;;) {
+      for(i = 0;i < VFS_FD_SET_NUM;i++) {
+          if (vfs_fd_set[i].vfs) {
+              vfs_fd_set[i].test_readset   = vfs_fd_set[i].readset;
+              vfs_fd_set[i].test_writeset  = vfs_fd_set[i].writeset;
+              vfs_fd_set[i].test_exceptset = vfs_fd_set[i].exceptset;
+
+              CHECK_AND_CALL(ret, r, vfs_fd_set[i].vfs, select, maxfdp1, &vfs_fd_set[i].test_readset, &vfs_fd_set[i].test_writeset, &vfs_fd_set[i].test_exceptset, &fs_timeout);
+              if (ret < 0) {
+                  return ret;
+              }
+
+              if (ret >= 0) {
+                  num = num + ret;
+
+                  min_fd = vfs_fd_set[i].vfs->offset << VFS_INDEX_S;
+                  max_fd = (vfs_fd_set[i].vfs->offset + 1) << VFS_INDEX_S;
+
+                  for(fd = min_fd;fd < max_fd;fd++) {
+                      if (readset && !FD_ISSET(fd, &vfs_fd_set[i].test_readset)) {
+                          FD_CLR(fd, readset);
+                      }
+
+                      if (writeset && !FD_ISSET(fd, &vfs_fd_set[i].test_writeset)) {
+                          FD_CLR(fd, writeset);
+                      }
+
+                      if (exceptset && !FD_ISSET(fd, &vfs_fd_set[i].test_exceptset)) {
+                          FD_CLR(fd, exceptset);
+                      }
+                  }
+              }
+          }
+     }
+
+   // TODO: find a more elegat solution to stop select ....
+   if (__select_cancelled) {
+        __select_cancelled = 0;
+        __errno_r(r) = 4;
+        return -1;
+   }
+
+      // If some file descriptor are set, terminate scan
+      if (num > 0) {
+          break;
+      }
+
+      // Check timeout
+      if (msectimeout != 0) {
+          gettimeofday(&now, NULL);
+          if ((now.tv_sec - start.tv_sec) * 1000 - (((now.tv_usec - start.tv_usec) + 500) / 1000) >= msectimeout) {
+              break;
+          }
+      } else {
+         // If not timeout, exit
+         break;
+      }
+   }
+
+   return num;
+}
