--- /Users/jaumeolivepetrus/esp-idf/components/esp32/ld/esp32.common.ld	2017-01-09 15:53:18.000000000 +0100
+++ /Users/jaumeolivepetrus/Lua-RTOS-ESP32/ld/lua-rtos.ld	2017-01-09 15:48:30.000000000 +0100
@@ -163,6 +163,24 @@
     *(.xt_except_desc_end)
     *(.dynamic)
     *(.gnu.version_d)
+
+	/* Lua RTOS changes (begin) */
+	
+    . = ALIGN(4);
+    lua_libs1 = ABSOLUTE(.);
+    KEEP (*(.lua_libs1))
+    LONG(0) LONG(0)
+
+    lua_rotable = ABSOLUTE(.);
+    KEEP(*(.lua_rotable1))
+    LONG(0) LONG(0)
+    
+    drivers = ABSOLUTE(.);
+    KEEP(*(.drivers))
+    LONG(0) LONG(0)
+
+	/* Lua RTOS changes (end) */
+
     _rodata_end = ABSOLUTE(.);
     /* Literals are also RO data. */
     _lit4_start = ABSOLUTE(.);
@@ -185,4 +200,4 @@
     _text_end = ABSOLUTE(.);
     _etext = .;
   } >iram0_2_seg
-}
+}
\ No newline at end of file
