diff --git a/components/nghttp/port/include/http_parser.h b/components/nghttp/port/include/http_parser.h
index 105ae510a..c3bf0e495 100644
--- a/components/nghttp/port/include/http_parser.h
+++ b/components/nghttp/port/include/http_parser.h
@@ -145,7 +145,7 @@ enum http_parser_type { HTTP_REQUEST, HTTP_RESPONSE, HTTP_BOTH };
 
 
 /* Flag values for http_parser.flags field */
-enum flags
+enum parser_flags
   { F_CHUNKED               = 1 << 0
   , F_CONNECTION_KEEP_ALIVE = 1 << 1
   , F_CONNECTION_CLOSE      = 1 << 2
