diff --git a/components/lwip/port/freertos/sys_arch.c b/components/lwip/port/freertos/sys_arch.c
index 2c56969a..97b986c0 100755
--- a/components/lwip/port/freertos/sys_arch.c
+++ b/components/lwip/port/freertos/sys_arch.c
@@ -449,8 +449,6 @@ sys_init(void)
 
     // Create the pthreads key for the per-thread semaphore storage
     pthread_key_create(&sys_thread_sem_key, sys_thread_sem_free);
-
-    esp_vfs_lwip_sockets_register();
 }
 
 /*-----------------------------------------------------------------------------------*/
